File: CPP/Stack/StockSpan.cpp
#include <bits/stdc++.h> 
using namespace std; 

vector<int> stockSpan(int arr[], int n) {
	stack<int> s;
	vector<int> ans;
	vector<int> span;

	for(int i = 0; i < n; i++) {
	    while(!s.empty() && arr[i] >= arr[s.top()]) {
			s.pop();
		}
		if(s.empty()) {
		    ans.push_back(-1);
		}
		else  {
		    ans.push_back(s.top());
		}
		s.push(i);
	}
	for(int i = 0; i < n; i++) {
		span.push_back(i - ans[i]);
	}
	return span;
}

int main(int argc, char const *argv[]) {
	
	int arr[] = {100, 80, 60, 70, 60, 75, 85}; 
	int n = sizeof(arr) / sizeof(arr[0]); 
	vector<int> ans = stockSpan(arr, n); 
	for (int a: ans) {
		cout << a << " ";
	}

}

File: Python/Stack/StockSpan.py
class StockSpan:

    def stockSpan(arr, n):
        stack = []
        ans = []
        span = []

        for i in range(n):
            while stack and arr[i] >= arr[stack[-1]]:
                stack.pop()
            if len(stack) == 0:
                ans.append(-1)
            else:
                ans.append(stack[-1])
            stack.append(i)

        for i in range(n):
            span.append(i - ans[i])
        return span

if __name__ == '__main__':
    arr = [100, 80, 60, 70, 60, 75, 85]
    ans = StockSpan.stockSpan(arr, len(arr))
    print(ans)